webhooks/virtualmachine/validation/virtualmachine_validator.go

```diff
@@ -475,6 +471,98 @@ func (v validator) validateReadinessProbe(ctx *context.WebhookRequestContext, vm
 	return allErrs
 }
 
+func (v validator) validatePowerStateOnCreate(
+	ctx *context.WebhookRequestContext,
+	newVM *vmopv1.VirtualMachine) field.ErrorList {
+
+	var (
+		allErrs       field.ErrorList
+		newPowerState = newVM.Spec.PowerState
+	)
+
+	if newPowerState == vmopv1.VirtualMachineSuspended {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				field.NewPath("spec").Child("powerState"),
+				newPowerState,
+				fmt.Sprintf(invalidPowerStateOnCreateFmt, newPowerState)))
+	}
+
+	return allErrs
+}
+
+func (v validator) validatePowerStateOnUpdate(
+	ctx *context.WebhookRequestContext,
+	newVM, oldVM *vmopv1.VirtualMachine) field.ErrorList {
+
+	var allErrs field.ErrorList
+	powerStatePath := field.NewPath("spec").Child("powerState")
+
+	// If a VM is powered off, all config changes are allowed except for setting
+	// the power state to shutdown or suspend.
+	// If a VM is requesting a power off, we can Reconfigure the VM _after_
+	// we power it off - all changes are allowed.
+	// If a VM is requesting a power on, we can Reconfigure the VM _before_
+	// we power it on - all changes are allowed.
+	newPowerState, oldPowerState := newVM.Spec.PowerState, oldVM.Spec.PowerState
+
+	if newPowerState == "" {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				powerStatePath,
+				newPowerState,
+				invalidPowerStateOnUpdateEmptyString))
+		return allErrs
+	}
+
+	switch oldPowerState {
+	case vmopv1.VirtualMachinePoweredOn:
+		// The request is attempting to power on a VM that is already powered
+		// on. Validate fields that may or may not be mutable while a VM is in
+		// a powered on state.
+		if newPowerState == vmopv1.VirtualMachinePoweredOn {
+			invalidFields := v.validateUpdatesWhenPoweredOn(ctx, newVM, oldVM)
+			allErrs = append(allErrs, invalidFields...)
+		}
+
+	case vmopv1.VirtualMachinePoweredOff:
+		// Mark the power state as invalid if the request is attempting to
+		// gracefully shutdown or suspend a VM that is powered off.
+		var msg string
+		if newPowerState == vmopv1.VirtualMachineSuspended {
+			msg = "suspend"
+		}
+		if msg != "" {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					powerStatePath,
+					newPowerState,
+					fmt.Sprintf(invalidPowerStateOnUpdateFmt, msg, "powered off")))
+		}
+
+	case vmopv1.VirtualMachineSuspended:
+		// Mark the power state as invalid if the request is attempting to
+		// halt or shutdown a VM that is suspended.
+		var msg string
+		if newPowerState == vmopv1.VirtualMachinePoweredOff {
```

Comment:
I believe Suspended -> PowerOff is a supported transition (my local ESXI allows it), and it seems reasonable if you cannot power on the VM due to like resource constraints (or just don't want to) power off should still be an option to free up the suspend state.

Reply:
Ack. I guess it just acts like a halt on a suspended physical system.